File: src/data/steer_command.cpp
#include "data/steer_command.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Data
{
SteerCommand::SteerCommand(const PbData::SteerCommand& protobuf) {
    angleDegrees = protobuf.angledegrees();
}

SteerCommand::operator PbData::SteerCommand() const {
    PbData::SteerCommand ret;
    ret.set_angledegrees(angleDegrees);
    return ret;
}

std::string SteerCommand::serializeAsJsonString() const {
    PbData::SteerCommand protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string SteerCommand::serializeAsProtobufString() const {
    PbData::SteerCommand protobuf(*this);
    return protobuf.SerializeAsString();
}

bool SteerCommand::deserializeFromJsonString(const std::string& str) {
    PbData::SteerCommand protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool SteerCommand::deserializeFromProtobufString(const std::string& str) {
    PbData::SteerCommand protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

SteerStatus::SteerStatus(const PbData::SteerStatus& protobuf) {
    status = static_cast<Status>(protobuf.status());
}

SteerStatus::operator PbData::SteerStatus() const {
    PbData::SteerStatus ret;
    ret.set_status(static_cast<PbData::Status>(status));
    return ret;
}

std::string SteerStatus::serializeAsJsonString() const {
    PbData::SteerStatus protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string SteerStatus::serializeAsProtobufString() const {
    PbData::SteerStatus protobuf(*this);
    return protobuf.SerializeAsString();
}

bool SteerStatus::deserializeFromJsonString(const std::string& str) {
    PbData::SteerStatus protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool SteerStatus::deserializeFromProtobufString(const std::string& str) {
    PbData::SteerStatus protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/data/steer_command.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from data import steer_command_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class SteerCommand:
    angleDegrees: float = 0.0
    
    _proto_message: steer_command_pb2.SteerCommand = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = steer_command_pb2.SteerCommand()

    def _populate_proto(self):
        self._proto_message.angleDegrees = self.angleDegrees

    @classmethod
    def _from_proto(cls, proto_message) -> "SteerCommand":
        return cls(
            angleDegrees = proto_message.angleDegrees,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "SteerCommand":
        message = steer_command_pb2.SteerCommand()
        message.ParseFromString(data)
        return cls(
            angleDegrees = message.angleDegrees,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "SteerCommand":
        message = steer_command_pb2.SteerCommand()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
class Status(Enum):
    steer_status_disabled = 0
    steer_status_enabled = 1


@dataclass
class SteerStatus:
    status: Status = None
    
    _proto_message: steer_command_pb2.SteerStatus = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = steer_command_pb2.SteerStatus()

    def _populate_proto(self):
        self._proto_message.status = self.status.value

    @classmethod
    def _from_proto(cls, proto_message) -> "SteerStatus":
        return cls(
            status = Status._from_proto(proto_message.status),
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "SteerStatus":
        message = steer_command_pb2.SteerStatus()
        message.ParseFromString(data)
        return cls(
            status = Status(message.status),
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "SteerStatus":
        message = steer_command_pb2.SteerStatus()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
